src/main/java/io/cryostat/net/openshift/OpenShiftSecurityContext.java

```diff
@@ -0,0 +1,162 @@
+ * Works (as defined below), to deal in both
+ *
+ * (a) the Software, and
+ * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+ * one is included with the Software (each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.net.openshift;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import io.cryostat.net.security.SecurityContext;
+import io.cryostat.platform.ServiceRef;
+import io.cryostat.platform.ServiceRef.AnnotationKey;
+import io.cryostat.platform.discovery.AbstractNode;
+import io.cryostat.platform.discovery.BaseNodeType;
+import io.cryostat.platform.discovery.EnvironmentNode;
+import io.cryostat.platform.discovery.TargetNode;
+
+import dagger.Lazy;
+
+class OpenShiftSecurityContext implements SecurityContext {
+
+    static final String KEY_NAMESPACE = "SC_NAMESPACE";
+
+    private final Map<String, String> ctx = new HashMap<>();
+
+    OpenShiftSecurityContext(String namespace) {
+        this.ctx.put(KEY_NAMESPACE, namespace);
+    }
+
+    OpenShiftSecurityContext(Lazy<String> defaultNamespace, AbstractNode node) {
+        String ns;
+        ServiceRef serviceRef = findServiceRef(node);
+        if (serviceRef != null) {
+            ns = serviceRef.getCryostatAnnotations().get(AnnotationKey.NAMESPACE);
+            if (ns == null) {
+                // FIXME log properly
+                System.err.println(
+                        String.format(
+                                "ServiceRef [%s] did not have an annotations.cryostat.NAMESPACE"
+                                        + " value",
+                                serviceRef.getServiceUri()));
+                ns = defaultNamespace.get();
+            }
+        } else {
+            // FIXME
```

Comment:
Or I guess implement two-way links like you said in the previous thread?

Reply:
In the usual scenario this node is something below the Realm level in the tree, so in OpenShift that's a Namespace or anything below it, or it's something else provided by another plugin. If that Realm's subtree is well-formed then every path from the given node down should terminate in a `TargetNode`.

So this `else` clause handles the case where the subtree was not well-formed or some state update happened while we were processing the request, and the node we are looking at doesn't appear to be on some path between the Realm and a leaf node. If that is the case then the request seems to be invalid and should probably be failed early. If this is a persistent state and the given node really does exist but is not on a path that terminates in a target node leaf then it's just a corrupt state and not this class' responsibility to sort out what a sensible context is, and better to just fail so the user/admin can take a look and figure out what to do.

Otherwise, the only case where this `else` is entered when the tree is well-formed is if the given node is something between the root and the realm. In that situation, the user is trying to perform an action on essentially everything that Cryostat can see, not something within a specific namespace. We can either define this to be invalid and fail such requests, or we can allow it but require the user to have elevated cluster-wide privileges to do it, which is something else I touched on earlier when talking about the considerations for Automated Rules.